### PR TITLE
Refactor `inverse_eltype` calculations to use types.

### LINKED ISSUE #73: Inverse transformation for empty vectors fails
I am working on an algorithm where I want to define a transformation for a vector of length `p`. Importantly, `p` may be zero. This transformation is part of a tuple of transformations. Therefore, the final full parameter vector has a non-zero dimension. What I have done in the past is to omit the zero-dimensional transform, but this leads to additional coding to "compensate" for the missing element. Also, I think it can result in type instability.

I can define a transform for a vector of dimension zero. Also, the forward transform does work:
```julia
julia> trans = as(Vector,0)
TransformVariables.ArrayTransform{TransformVariables.Identity,1}(asℝ, (0,))
julia> trans([])
0-element Array{Any,1}
```
However, the inverse transform fails:
```julia
julia> inverse(trans,[])
ERROR: BoundsError: attempt to access 0-element Array{Any,1} at index [1]
Stacktrace:
 [1] getindex at ./array.jl:788 [inlined]
 [2] first at ./abstractarray.jl:323 [inlined]
 [3] inverse_eltype(::TransformVariables.ArrayTransform{TransformVariables.Identity,1}, ::Array{Any,1}) at /home/student/.julia/packages/TransformVariables/gAt5r/src/aggregation.jl:78
 [4] inverse(::TransformVariables.ArrayTransform{TransformVariables.Identity,1}, ::Array{Any,1}) at /home/student/.julia/packages/TransformVariables/gAt5r/src/generic.jl:236
 [5] top-level scope at REPL[47]:1
 [6] eval(::Module, ::Any) at ./boot.jl:331
 [7] eval_user_input(::Any, ::REPL.REPLBackend) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.4/REPL/src/REPL.jl:86
 [8] run_backend(::REPL.REPLBackend) at /home/student/.julia/packages/Revise/Pcs5V/src/Revise.jl:1073
 [9] top-level scope at none:0
```
It would be great if this could work.

Just as a thought - I am not sure if this makes sense in the grand scheme of things - but perhaps this and other cases could be handled by having the option to define the transform for a constant? This would also have a zero-dimensional parameter space, and zero jacobian.

### PR DESCRIPTION
**Pull request by tpapp**
This is a continuation of #133 in a sense, using the approach to fix some other issues.

Incidental changes:

- whitespace and formatting
- clarify that inverse_eltype does (may) not do input checks
- fixed #73, the corner case of the empty vector; tests added
- rename utility function introduced in #133 and place in utilities.jl

### CODE REVIEW

**Review by devmotion — COMMENTED** [review]
I'll review in more detail later today or tomorrow. Generally, it looks fine but this is a breaking change (`inverse_eltype` is part of the documented interface and with this PR `inverse_eltype` of `TransformTuple` etc call the definition with types which mostly likely downstream packages do not define currently and hence will throw an error).

File: test/runtests.jl [hunk 1]
```diff
@@ -840,3 +850,11 @@ end
...
     end
 end
+
+####
+#### static analysis with JET
+####
+
+import JET
+@test isempty(JET.get_reports(JET.report_package(TransformVariables;
+                                                 target_modules = (TransformVariables,))))
```
Comment (devmotion) [inline test/runtests.jl:860]: A bit simpler (but equivalent IIRC) would be
```suggestion
JET.test_package(TransformVariables; target_modules = (TransformVariables,))
```

File: src/aggregation.jl [hunk 2]
```diff
@@ -341,8 +345,17 @@ internally.
...
+    ex = mapreduce((i, T) -> :(inverse_eltype(ts[$i], $T)),
+                   (a, b) -> :(promote_type($a, $b)),
+                   1:N, fieldtypes(T))
+    quote
+        A = ts
+        B = fieldtypes(T)
+        @argcheck N == length(B) "Incompatible input length."
+        $(ex)
+    end
+end
```
Comment (devmotion) [inline src/aggregation.jl:358]: I'm not a big fan of generated functions (but sometimes it's unavoidable) - I wonder if a recursive definition is inferred as well, something like
```suggestion
function _inverse_eltype_tuple(ts::NTransforms{N}, ::Type{T}) where {N,T<:Tuple}
    @argcheck T <: NTuple{N,Any} "Incompatible input length."
    return __inverse_eltype_tuple(ts, T)
end
function __inverse_eltype_tuple(ts::NTransform, ::Type{Tuple{}})
   return Union{}
end
function __inverse_eltype_tuple(ts::NTransform, ::Type{T}) where {T<:Tuple}
   return promote_type(inverse_eltype(Base.first(ts), fieldtype(T, 1)), __inverse_eltype_tuple(Base.tail(ts), Tuple{Base.tail(fieldtypes(T))...}))
end
```

**Review by devmotion — APPROVED** [review]

**Comment by tpapp** [comment]
I decided pull back the 1.0.0 release, release this as 0.9.0, and fix other breaking changes before 1.0.0.

@stjordanis, I am pinging you here since the above Pumas.jl repo does not accept issues.

**Comment by devmotion** [comment]
You don't have to worry about that repo, it's a completely outdated version of Pumas that was copied by the Github user when its source code was publicly available.
